util/logger.go

```diff
@@ -1,55 +1,277 @@
+
 package util
 
 import (
-	"log"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
 	"os"
+	"strings"
 	"time"
+
+	"github.com/rs/zerolog"
 )
 
-type LogLevel int
+var Log *LoggerService
 
-const (
-	INFO LogLevel = iota
-	WARN
-	ERROR
-)
+type LoggerService struct {
+	Logger zerolog.Logger
+	Env    string
+}
+
+type ILoggerService interface {
+	// Function to enrich each log with data
+	enrich(req *http.Request, e *zerolog.Event) *zerolog.Event
+
+	// This set of functions is to be used in the context of the web-server
+	// where there is a server context involved
+	DebugCtx(req *http.Request, msg string)
+	InfoCtx(req *http.Request, msg string)
+	WarnCtx(req *http.Request, msg string)
+	ErrorCtx(req *http.Request, msg string, err error)
+	FatalCtx(req *http.Request, msg string, err error)
+	PanicCtx(req *http.Request, msg string, r any, trace string) // r = recover()
+	SuccessCtx(req *http.Request)
+
+	// This set of functions can be used in scenarios where there is no
+	// server context involved
+	Debug(msg string)
+	Info(msg string)
+	Warn(msg string)
+	Error(msg string, err error)
+	Fatal(msg string, err error)
 
-type Logger struct {
-	logger *log.Logger
-	level  LogLevel
+	// Logging middleware to be used only as a global middleware during router
+	// initialization
+	LogMiddleware(next http.Handler) http.Handler
 }
 
-func NewLogger() *Logger {
-	return &Logger{
-		logger: log.New(os.Stdout, "", 0),
-		level:  INFO,
+func InitLogger(env string) (*LoggerService, error) {
+	var output io.Writer
+
+	switch env {
+	case "DEVELOPMENT":
+		file, err := os.OpenFile("dev.log", os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0666)
+		if err != nil {
+			return nil, err
+		}
+		consoleWriter := zerolog.ConsoleWriter{
+			Out:        os.Stderr,
+			TimeFormat: time.RFC3339,
+		}
+		fileWriter := zerolog.ConsoleWriter{
+			Out:        file,
+			TimeFormat: "",
+			FormatFieldName: func(i any) string {
+				return fmt.Sprintf("%s=", i)
+			},
+			FormatFieldValue: func(i any) string {
+				s := fmt.Sprintf("%v", i)
+				if strings.ContainsAny(s, "") {
+					return fmt.Sprintf("%q", s)
+				}
+				return s
+			},
+			FormatTimestamp: func(i any) string {
+				t, err := time.Parse(time.RFC3339, i.(string))
+				if err != nil {
+					return fmt.Sprintf("time=%q", i) // Fallback if parsing fails
+				}
+				return fmt.Sprintf("time=%d", t.UnixMilli())
+			},
+			FormatLevel: func(i any) string {
+				return fmt.Sprintf("level=%q", i)
+			},
+			FormatMessage: func(i any) string {
+				return fmt.Sprintf("msg=%q", i) // Quoting the message automatically
+			},
+			NoColor: true,
+		}
+		output = zerolog.MultiLevelWriter(consoleWriter, fileWriter)
+	case "PRODUCTION":
```

Comment:
Define const values such as `PRODUCTION` as Enums(using const in go). It will be awesome if the values are set from an env variable.